core/primitives/src/sharding.rs

```diff
@@ -404,16 +427,36 @@ impl ShardChunkHeader {
         }
     }
 
+    #[inline]
+    pub fn congestion_info(&self) -> Option<CongestionInfo> {
+        match self {
+            ShardChunkHeader::V1(_) => None,
+            ShardChunkHeader::V2(_) => None,
+            ShardChunkHeader::V3(header) => header.inner.congestion_info(),
+        }
+    }
+
     /// Returns whether the header is valid for given `ProtocolVersion`.
     pub fn valid_for(&self, version: ProtocolVersion) -> bool {
         const BLOCK_HEADER_V3_VERSION: ProtocolVersion =
             ProtocolFeature::BlockHeaderV3.protocol_version();
+        const CONGESTION_CONTROL_VERSION: ProtocolVersion =
+            ProtocolFeature::CongestionControl.protocol_version();
+
         match &self {
             ShardChunkHeader::V1(_) => version < SHARD_CHUNK_HEADER_UPGRADE_VERSION,
             ShardChunkHeader::V2(_) => {
                 SHARD_CHUNK_HEADER_UPGRADE_VERSION <= version && version < BLOCK_HEADER_V3_VERSION
             }
-            ShardChunkHeader::V3(_) => BLOCK_HEADER_V3_VERSION <= version,
+            ShardChunkHeader::V3(header) => match header.inner {
+                ShardChunkHeaderInner::V1(_) => {
+                    version >= BLOCK_HEADER_V3_VERSION && version < CONGESTION_CONTROL_VERSION
+                }
+                ShardChunkHeaderInner::V2(_) => {
+                    version >= BLOCK_HEADER_V3_VERSION && version < CONGESTION_CONTROL_VERSION
+                }
+                ShardChunkHeaderInner::V3(_) => version >= CONGESTION_CONTROL_VERSION,
+            },
```

Comment:
Looks correct, but I gotta say, this version handling is getting is a bit insane :S
But not your fault ;)

Nit: Personally, I find it easier to read when cases with identical code are summarized. But feel free to keep at is if you like them separate.

```suggestion
            ShardChunkHeader::V3(header) => match header.inner {
                ShardChunkHeaderInner::V1(_) | ShardChunkHeaderInner::V2(_) => {
                    version >= BLOCK_HEADER_V3_VERSION && version < CONGESTION_CONTROL_VERSION
                }
                ShardChunkHeaderInner::V3(_) => version >= CONGESTION_CONTROL_VERSION,
            },
```